libursa/benches/bench_wrappers/signatures.rs

```diff
@@ -0,0 +1,204 @@
+        assert!(result.is_ok());
+    }
+
+    pub fn load_keys_benchmark(c: &mut Criterion) {
+        c.bench_function("ed25519 key loading", |b| b.iter(|| load_keys()));
+    }
+
+    fn sign() {
+        let scheme = Ed25519Sha512::new();
+        let s = PrivateKey(hex2bin(PRIVATE_KEY).unwrap());
+        let result = scheme.sign(MESSAGE_1, &s);
+        assert!(result.is_ok());
+    }
+
+    pub fn sign_benchmark(c: &mut Criterion) {
+        c.bench_function("ed25519 signing", |b| b.iter(|| sign()));
+    }
+
+    fn verify() {
+        let scheme = Ed25519Sha512::new();
+        let p = PublicKey(hex2bin(PUBLIC_KEY).unwrap());
+        let result = scheme.verify(&MESSAGE_1, hex2bin(SIGNATURE_1).unwrap().as_slice(), &p);
+        assert!(result.is_ok());
+    }
+
+    pub fn verify_benchmark(c: &mut Criterion) {
+        c.bench_function("ed25519 verification", |b| b.iter(|| verify()));
+    }
+}
+
+pub mod secp256k1 {
+
+    use criterion::Criterion;
+
+    use ursa::encoding::hex;
+    use ursa::keys::{PrivateKey, PublicKey, KeyGenOption};
+    use ursa::signatures::*;
+    use ursa::signatures::secp256k1::EcdsaSecp256k1Sha256;
+    use libsecp256k1;
+    use ursa::sha2::Digest;
+    use openssl::ecdsa::EcdsaSig;
+    use openssl::ec::{EcGroup, EcPoint, EcKey};
+    use openssl::nid::Nid;
+    use openssl::bn::{BigNum, BigNumContext};
+
+    const MESSAGE_1: &[u8] = b"This is a dummy message for use with tests";
+    const SIGNATURE_1: &str = "ae46d3fec8e2eb95ebeaf95f7f096ec4bf517f5ef898e4379651f8af8e209ed75f3c47156445d6687a5f817fb3e188e2a76df653b330df859ec47579c8c409be";
+    const PRIVATE_KEY: &str = "e4f21b38e005d4f895a29e84948d7cc83eac79041aeb644ee4fab8d9da42f713";
+    const PUBLIC_KEY: &str = "0242c1e1f775237a26da4fd51b8d75ee2709711f6e90303e511169a324ef0789c0";
+
+    fn create_keys() {
+        let scheme = EcdsaSecp256k1Sha256::new();
+        let result = scheme.keypair(None);
+        assert!(result.is_ok());
+    }
+
+    pub fn create_keys_benchmark(c: &mut Criterion) {
+        c.bench_function("secp256k1 key creation", |b| b.iter(|| create_keys()));
+    }
+
+    fn load_keys() {
+        let scheme = EcdsaSecp256k1Sha256::new();
+        let secret = PrivateKey(hex::hex2bin(PRIVATE_KEY).unwrap());
+        let result = scheme.keypair(Some(KeyGenOption::FromSecretKey(secret)));
+        assert!(result.is_ok());
+    }
+
+    pub fn load_keys_benchmark(c: &mut Criterion) {
+        c.bench_function("secp256k1 key loading", |b| b.iter(|| load_keys()));
+    }
+
+    fn sign() {
+        let scheme = EcdsaSecp256k1Sha256::new();
+        let s = PrivateKey(hex::hex2bin(PRIVATE_KEY).unwrap());
+        let result = scheme.sign(MESSAGE_1, &s);
+        assert!(result.is_ok());
+    }
+
+    pub fn sign_benchmark(c: &mut Criterion) {
+        c.bench_function("secp256k1 signing", |b| b.iter(|| sign()));
+    }
+
+    fn sign_libsecp256k1() {
+        let context = libsecp256k1::Secp256k1::new();
```

Comment:
Can you please provide the benchmark without creating `context` each time?

Reply:
Memory allocation is usually a big performance problem in crypto implementations, especially on small payloads and in benchmarks. Moving these outside the implementation, or in this case, outside of the loop should help remove that extra noise/overhead.